File: projects/07-Treaps/Implementation/Treap.java
import java.util.*;

public class Treap {

    // Each node holds a key, a priority (heap), and two children (BST)
    private static final class Node {
        int key;
        double priority;
        Node left, right;

        Node(int key, Double pri, Random rnd) {
            this.key = key;
            this.priority = (pri != null) ? pri : rnd.nextDouble();
        }
    }

    // Split result holds two disjoint treap roots: left and right
    private static final class Split {
        Node left, right;
        Split(Node l, Node r) { left = l; right = r; }
    }

    private Node root;                     // root of the treap
    private final Random rnd = new Random(); // random priority generator


    // Restores heap order by rotating a node with its left child
    //      y              x
    //      / \            / \
    //     x   a   =>     c   y
    //    / \                / \
    //   c   b              b   a
    private Node rotateRight(Node y) {
        Node x = y.left;
        y.left = x.right;
        x.right = y;
        return x;
    }

    // Restores heap order by rotating a node with its right child
    //     x                  y
    //    / \                / \
    //   c   y    =>       x   a
    //      / \           / \
    //     b   a         c   b
    private Node rotateLeft(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }


    // Standard BST insert, then perform rotations to preserve heap property
    private Node insert(Node n, int key, Double pri) {
        if (n == null)
            return new Node(key, pri, rnd);  // reached leaf, create new node

        if (key < n.key) {
            n.left = insert(n.left, key, pri);   // insert to left subtree
            if (n.left.priority < n.priority)
                n = rotateRight(n);              // fix heap violation

        } else if (key > n.key) {
            n.right = insert(n.right, key, pri); // insert to right subtree
            if (n.right.priority < n.priority)
                n = rotateLeft(n);               // fix heap violation
        }
        // Ignore duplicate keys
        return n;
    }

    public void insert(int key) {
        root = insert(root, key, null); // use random priority
    }

    public void insert(int key, double pri) {
        root = insert(root, key, pri); // use user-supplied priority
    }


    // Find node by key, then rotate its children until leaf, then remove
    private Node delete(Node n, int key) {
        if (n == null) return null; // key not found

        if (key < n.key) {
            n.left = delete(n.left, key); // search left

        } else if (key > n.key) {
            n.right = delete(n.right, key); // search right

        } else {
            // node found
            // to safely remove the node, need to rotate it down until it becomes a leaf
            if (n.left == null || n.right == null)
                return (n.left != null) ? n.left : n.right; // 0 or 1 child

            // rotate smaller priority child up and repeat delete
            if (n.left.priority < n.right.priority) {
                n = rotateRight(n);
                n.right = delete(n.right, key);
            } else {
                n = rotateLeft(n);
                n.left = delete(n.left, key);
            }
        }
        return n;
    }

    public void delete(int key) {
        root = delete(root, key);
    }


    // Standard BST search: returns true if key exists
    public boolean search(int key) {
        Node cur = root;
        while (cur != null) {
            if (key < cur.key) cur = cur.left;      // go left
            else if (key > cur.key) cur = cur.right; // go right
            else return true;                        // key found
        }
        return false; // not found
    }

    // Recursively split treap into (< key) and (>= key)

    private Split split(Node n, int key) {
        if (n == null) return new Split(null, null); // base case

        if (key <= n.key) {
            Split s = split(n.left, key);  // recurse to the left subtree to find the correct split point
            n.left = s.right;              // reattach right to current node
            return new Split(s.left, n);
        } else {
            Split s = split(n.right, key); // recurse right
            n.right = s.left;              // reattach left to current node
            return new Split(n, s.right);
        }
    }

    public Treap[] split(int key) {
        Split s = split(root, key);        // split root treap
        Treap left = new Treap();
        Treap right = new Treap();
        left.root = s.left;                // assign split results
        right.root = s.right;
        return new Treap[] { left, right };
    }


    // Combine two treaps assuming all keys in L < all keys in R
    private Node merge(Node L, Node R) {
        if (L == null || R == null)
            return (L != null) ? L : R; // base case

        if (L.priority < R.priority) {
            // 'left' has higher priority (remember: smaller number = higher priority)
            // So it becomes the new root of the merged treap.
            // Recursively merge its right subtree with 'right', and reattach.
            L.right = merge(L.right, R);  // recurse right on L
            return L;
        } else {
            // 'right' has higher priority, so it becomes the new root.
            // Recursively merge 'left' with its left subtree, and reattach.
            R.left = merge(L, R.left);    // recurse left on R
            return R;
        }
    }

    public static Treap merge(Treap left, Treap right) {
        Treap t = new Treap();
        t.root = t.merge(left.root, right.root); // call internal merge
        return t;
    }


    // Iterative inorder traversal: returns sorted list of keys
    public List<Integer> inorder() {
        List<Integer> result = new ArrayList<>();
        Deque<Node> stack = new ArrayDeque<>();
        Node cur = root;

        while (cur != null || !stack.isEmpty()) {
            while (cur != null) {
                stack.push(cur);       // go as far left as possible
                cur = cur.left;
            }
            cur = stack.pop();         // visit node
            result.add(cur.key);       // add key to result
            cur = cur.right;           // process right child
        }

        return result;
    }



    public static void main(String[] args) throws Exception {
        Treap t = new Treap();
        List<String> lines = java.nio.file.Files.readAllLines(java.nio.file.Paths.get(args[0]));

        String mode = lines.get(0).trim();
        String argsLine = lines.get(1).trim();

        if (mode.equals("Merge")) {
            String[] parts = argsLine.split(" ");
            int n1 = Integer.parseInt(parts[0]);
            int n2 = Integer.parseInt(parts[1]);
            List<String> ops1 = lines.subList(2, 2 + n1);
            List<String> ops2 = lines.subList(2 + n1, 2 + n1 + n2);
            Treap t1 = runTreapOps(ops1);
            Treap t2 = runTreapOps(ops2);
            Treap merged = Treap.merge(t1, t2);
            System.out.println("Merged inorder: " + merged.inorder());

        } else if (mode.equals("Split")) {
            int n = Integer.parseInt(argsLine);
            List<String> ops = lines.subList(2, 2 + n);
            int splitKey = Integer.parseInt(lines.get(2 + n).trim());
            Treap tSplit = runTreapOps(ops);
            Treap[] parts = tSplit.split(splitKey);
            System.out.println("Left treap inorder: " + parts[0].inorder());
            System.out.println("Right treap inorder: " + parts[1].inorder());

        } else if (mode.equals("Basic")) {
            int n = Integer.parseInt(argsLine);
            List<String> ops = lines.subList(2, 2 + n);
            runTreapOps(ops);
        }
    }

    // Execute a list of operations like Insert/Delete/Search/Inorder
    private static Treap runTreapOps(List<String> lines) {
        Treap t = new Treap();
        for (String line : lines) {
            String[] parts = line.trim().split("\\s+");
            if (parts.length == 0) continue;
            String cmd = parts[0];
            if (cmd.equals("Insert")) {
                t.insert(Integer.parseInt(parts[1]));
            } else if (cmd.equals("Delete")) {
                t.delete(Integer.parseInt(parts[1]));
            } else if (cmd.equals("Search")) {
                boolean found = t.search(Integer.parseInt(parts[1]));
                System.out.println("Search " + parts[1] + ": " + (found ? "Found" : "Not Found"));
            } else if (cmd.equals("Inorder")) {
                System.out.println("Inorder: " + t.inorder());
            }
        }
        return t;
    }
}

File: projects/07-Treaps/Implementation/Treap.py
import random

# Node of treap: stores key, heap priority, and children
class Node:
    def __init__(self, key, priority=None):
        self.key = key
        self.priority = priority if priority is not None else random.random()
        self.left = None
        self.right = None


class Treap:
    def __init__(self):
        self.root = None

    # Right rotate to fix heap violation
    #       y              x
    #      / \            / \
    #     x   a   =>     c   y
    #    / \                / \
    #   c   b              b   a
    def _rotate_right(self, y):
        x = y.left
        y.left = x.right
        x.right = y
        return x

    # Left rotate to fix heap violation
    #     x                  y
    #    / \                / \
    #   c   y    =>       x   a
    #      / \           / \
    #     b   a         c   b
    def _rotate_left(self, x):
        y = x.right
        x.right = y.left
        y.left = x
        return y

    # Insert a key with optional priority
    # Standard BST insert, then rotate if heap violated
    def _insert(self, node, key, priority=None):
        if node is None:
            # Reached a null position, create new node
            return Node(key, priority)

        if key < node.key:
            node.left = self._insert(node.left, key, priority)
            if node.left.priority < node.priority:
                # If the left child has a higher priority (smaller number), rotate right
                node = self._rotate_right(node)

        elif key > node.key:
            node.right = self._insert(node.right, key, priority)
            if node.right.priority < node.priority:
                # If the right child has a higher priority, rotate left
                node = self._rotate_left(node)

        return node

    def insert(self, key, priority=None):
        self.root = self._insert(self.root, key, priority)

    # Delete key by BST search, then rotate to leaf
    def _delete(self, node, key):
        if node is None:
            return None

        if key < node.key:
            node.left = self._delete(node.left, key)
        elif key > node.key:
            node.right = self._delete(node.right, key)
        else:
            # Node found, now handle 0, 1, or 2 children cases
            # to safely remove the node, need to rotate it down until it becomes a leaf
            if node.left is None and node.right is None:
                # Case 1: Leaf node
                return None
            elif node.left and node.right:
                # Case 3: Two children → rotate child with smaller priority
                if node.left.priority < node.right.priority:
                    node = self._rotate_right(node)
                    node.right = self._delete(node.right, key)
                else:
                    node = self._rotate_left(node)
                    node.left = self._delete(node.left, key)
            else:
                # Case 2: One child → replace node with child
                node = node.left or node.right
        return node

    def delete(self, key):
        self.root = self._delete(self.root, key)

    # Standard BST search
    def search(self, key):
        cur = self.root
        while cur:
            if key < cur.key:
                cur = cur.left
            elif key > cur.key:
                cur = cur.right
            else:
                return True  # Key found
        return False

    # Split treap into two treaps based on key
    # Left treap has keys < key, right treap has keys >= key
    def _split(self, node, key):
        if node is None:
            return (None, None)

        if key <= node.key:
            #recurse to the left subtree to find the correct split point
            L, R = self._split(node.left, key)
            node.left = R  # Attach leftover right split to node
            return (L, node)
        else:
            L, R = self._split(node.right, key)
            node.right = L  # Attach leftover left split to node
            return (node, R)

    def split(self, key):
        L, R = self._split(self.root, key)
        left_t, right_t = Treap(), Treap()
        left_t.root, right_t.root = L, R
        return (left_t, right_t)

    # Merge two treaps: left < right, heap priority decides root
    def _merge_nodes(self, left, right):
        if left is None or right is None:
            return left or right  # One tree is empty

        if left.priority < right.priority:
            # 'left' has higher priority (remember: smaller number = higher priority)
            # So it becomes the new root of the merged treap.
            # Recursively merge its right subtree with 'right', and reattach.
            left.right = self._merge_nodes(left.right, right)
            return left
        else:
            # 'right' has higher priority, so it becomes the new root.
            # Recursively merge 'left' with its left subtree, and reattach.
            right.left = self._merge_nodes(left, right.left)
            return right

    @staticmethod
    def merge(left_treap: "Treap", right_treap: "Treap") -> "Treap":
        t = Treap()
        t.root = t._merge_nodes(left_treap.root, right_treap.root)
        return t

    # In-order traversal for sorted keys
    def inorder(self, node=None):
        if node is None:
            node = self.root
        res = []
        def _rec(n):
            if not n:
                return
            _rec(n.left)         # Traverse left subtree
            res.append(n.key)    # Visit node
            _rec(n.right)        # Traverse right subtree
        _rec(node)
        return res


# Parse and execute commands on a treap
def run_treap_ops(lines):
    t = Treap()
    for line in lines:
        parts = line.strip().split()
        if not parts:
            continue
        cmd = parts[0]
        if cmd == "Insert":
            t.insert(int(parts[1]))
        elif cmd == "Delete":
            t.delete(int(parts[1]))
        elif cmd == "Search":
            found = t.search(int(parts[1]))
            print(f"Search {parts[1]}: {'Found' if found else 'Not Found'}")
        elif cmd == "Inorder":
            print("Inorder:", t.inorder())
    return t

# Main function: read mode and input, execute treap ops
def main():
    import sys
    file_path = sys.argv[1]
    with open(file_path, "r") as f:
        lines = f.readlines()

    mode = lines[0].strip()
    args = lines[1].strip()

    if mode == "Merge":
        # Merge two separate treaps
        n1, n2 = map(int, args.split())
        ops1 = lines[2:2+n1]
        ops2 = lines[2+n1:2+n1+n2]
        t1 = run_treap_ops(ops1)
        t2 = run_treap_ops(ops2)
        merged = Treap.merge(t1, t2)
        print("Merged inorder:", merged.inorder())

    elif mode == "Split":
        # Split a single treap
        n = int(args)
        ops = lines[2:2+n]
        split_key = int(lines[2+n])
        t = run_treap_ops(ops)
        L, R = t.split(split_key)
        print("Left treap inorder:", L.inorder())
        print("Right treap inorder:", R.inorder())

    elif mode == "Basic":
        # Just basic operations
        n = int(args)
        ops = lines[2:2+n]
        t = run_treap_ops(ops)

if __name__ == "__main__":
    main()
